typing/ctype.ml

```diff
@@ -2589,6 +2596,20 @@ let record_equation t1 t2 =
   | Forbidden -> assert false
   | Allowed { equated_types } -> TypePairs.add equated_types (t1, t2) ()
 
+(* Called from unify3 *)
+let unify3_var env t1' t2 t2' =
+  occur !env t1' t2;
+  try
+    occur_univar !env t2;
+    link_type t1' t2;
+  with Unify _ when !umode = Pattern ->
+    reify env t1';
+    reify env t2';
+    if can_generate_equations () then begin
+      occur_univar ~inj_only:true !env t2';
```

Comment:
Why is `occur_univar ~inj_only:true` guarded by `can_generate_equations`? I would expect it to be safe to call that even when not generating equations.

Reply:
`can_generate_equations` is a bit of a misnomer. It means that we are in a non-injective branch, and failing will result in undoing unifications and reifying larger types, i.e. getting a less precise type. See the last case of `Tconstr _, _` in `unify3` for another example skipping a check (using `mcomp` instead of `occur_univar ~inj_only:true`).